documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md

```diff
@@ -6,283 +6,268 @@ Proposal design for a platform invoke (P/Invoke) mechanism for Silk.NET 3.0.
 
-Consider the following example:
-```cs
-#if __IOS__
-[UseStaticLibrary]
-#endif
-[UseDynamicLibrary("glfw3.dll")]
-public partial class Glfw
-{
-    public partial void glfwInit();
-}
-```
+Tanner has already stated that they're happy to work with us to get Silk.NET using ClangSharp, even saying they're happy to add a CI test stage into the ClangSharp repo to ensure no incoming changes break Silk.NET's generation process.
 
-The Emitter **MUST** only generate code if all preprocessor directives guarding the `UseStaticLibrary` attribute evaluate to true. If the attribute is defined on both the function and the containing type, the preprocessor directives surrounding the function's attribute **MUST** be used instead of the preprocessor directives surrounding the type.
+# Implicit Cast-Based Overloading
 
-Note to the reader: Given preprocessor directives are processed at parse time in Roslyn, both of those last requirements are basically benign.
+The pivotal proposal around which the SilkX experiment revolves is solving the excessive overload problem by having no overloads at all. Namely, Silk.NET.Core will have a set of "pointer-like" types that can be implicitly casted to by a multitude of different types. This gives the user the same flexibility as having all different permutations of overloaded parameters generated without having to do so, and thereby alleviates the requirement of having overloads at all.
 
-#### Custom Call Style: Procedure Address Expressions
+`Ptr` and `Mut` `ref struct`s **shall** be defined wrapping a `ref readonly` and `ref` respectively. These types **shall** have generic and non-generic variants. This shall be repeated up to 3 pointer dimensions, taking into account C's mutability rules in each instantiation. This means that there will be 14 top-level generic structs and 14 top-level non-generic structs. These **should** be auto-generated to ensure they remain maintainable. IL weaving **may** be required for the higher dimensions.
 
-Custom code may be used as a call style by providing a pointer to SilkTouch using a Procedure Address Expression. This is useful in scenarios such as COM interop.
+Unless deemed inappropriate, inapplicable, and/or infeasible by the Silk.NET team (the team reserved the right to do so without Working Group approval), the types **should** have the following characteristics:
+- An instance **must** be constructable from a `ref T` or `ref readonly T` (depending on the outer dimension mutability) where `T` is either the generic type (or `byte` for non-generic variants), a "pointer-like" type of the inner dimension of type being constructed.
```

Comment:
Regarding dimension mutability, more explicit definition maybe good. As of right now inner-most is first and outermost is last. This may also not be the most intuitive as I read it in the reverse. So MutPtr<byte> would be byte** const rather than byte * const *.